src/qcodes/dataset/data_set.py

```diff
@@ -251,9 +252,18 @@ def __init__(
                 Ignored if ``run_id`` is provided.
             in_memory_cache: Should measured data be keep in memory
                 and available as part of the `dataset.cache` object.
+            read_only: whether to open the connection in read-only mode.
+                Only takes effect if `conn` is not given.
 
         """
-        self.conn = conn_from_dbpath_or_conn(conn, path_to_db)
+        if run_id is None and read_only:
+            # raise valueerror here because if no run id, a new dataset will be created
+            # and it will be written to the database
+            raise ValueError(
```

Comment:
Could you add a test for this error? And also a test of where read_only flag does nothing in case connection instance is passed explicitly.